File: src/stripe_entitlements/credit_amount.py
from __future__ import annotations

import re
from dataclasses import dataclass
from decimal import Decimal
from typing import Any

from .bounds import POSTGRES_BIGINT_MAX

CREDIT_SCALE = 1_000_000
CREDIT_DECIMAL_PLACES = 6
MAX_CREDIT_ATOMS = POSTGRES_BIGINT_MAX
MAX_WHOLE_CREDITS = MAX_CREDIT_ATOMS // CREDIT_SCALE

_PLAIN_DECIMAL = re.compile(r"^(0|[1-9][0-9]*)(?:\.([0-9]{1,6}))?$")


def _plain_decimal_text(value: Any, *, field: str) -> str:
    if type(value) is int:
        if value < 0:
            raise ValueError(f"{field} must be non-negative")
        return str(value)
    if isinstance(value, Decimal):
        if not value.is_finite():
            raise ValueError(f"{field} must be finite")
        text = str(value)
        if "e" in text.lower():
            raise ValueError(f"{field} must not use exponent notation")
        return text
    if type(value) is str:
        return value
    if isinstance(value, (bool, float)):
        raise ValueError(f"{field} must not be a float or boolean")
    raise ValueError(f"{field} must be an integer, Decimal, or plain decimal string")


def credit_atoms(value: Any, *, field: str = "credit amount", allow_zero: bool = True) -> int:
    """Normalize a logical credit amount to exact fixed-point atoms.

    Integers retain the public API's historical meaning of whole credits. Decimal
    strings are deliberately parsed without binary floating-point or implicit
    rounding. Persisted values and internal arithmetic use the returned atoms.
    """

    text = _plain_decimal_text(value, field=field)
    match = _PLAIN_DECIMAL.fullmatch(text)
    if match is None:
        raise ValueError(
            f"{field} must be a non-negative plain decimal with at most "
            f"{CREDIT_DECIMAL_PLACES} fractional digits"
        )
    whole_text, fractional_text = match.group(1), match.group(2) or ""
    whole = int(whole_text)
    if whole > MAX_WHOLE_CREDITS:
        raise ValueError(f"{field} exceeds the PostgreSQL bigint atom range")
    atoms = whole * CREDIT_SCALE + int(fractional_text.ljust(CREDIT_DECIMAL_PLACES, "0") or 0)
    if atoms > MAX_CREDIT_ATOMS:
        raise ValueError(f"{field} exceeds the PostgreSQL bigint atom range")
    if not allow_zero and atoms == 0:
        raise ValueError(f"{field} must be greater than zero")
    return atoms


def credit_decimal(atoms: int, *, field: str = "credit atoms") -> str:
    """Serialize non-negative atoms as one canonical, exponent-free decimal string."""

    if type(atoms) is not int or atoms < 0 or atoms > MAX_CREDIT_ATOMS:
        raise ValueError(f"{field} must be a non-negative PostgreSQL bigint integer")
    whole, fractional = divmod(atoms, CREDIT_SCALE)
    if not fractional:
        return str(whole)
    return f"{whole}.{fractional:0{CREDIT_DECIMAL_PLACES}d}".rstrip("0")


def checked_add_atoms(left: int, right: int, *, field: str = "credit balance") -> int:
    """Add two non-negative atom quantities with an explicit bigint bound check."""

    if (
        type(left) is not int
        or type(right) is not int
        or left < 0
        or right < 0
        or left > MAX_CREDIT_ATOMS
        or right > MAX_CREDIT_ATOMS
    ):
        raise ValueError(f"{field} operands must be non-negative PostgreSQL bigint integers")
    if left > MAX_CREDIT_ATOMS - right:
        raise OverflowError(f"{field} exceeds the PostgreSQL bigint atom range")
    return left + right


@dataclass(frozen=True, slots=True, order=True)
class CreditAmount:
    """An exact, non-negative product-credit quantity stored as integer atoms."""

    atoms: int

    def __post_init__(self) -> None:
        if type(self.atoms) is not int or self.atoms < 0 or self.atoms > MAX_CREDIT_ATOMS:
            raise ValueError("credit atoms must be a non-negative PostgreSQL bigint integer")

    @classmethod
    def parse(
        cls,
        value: Any,
        *,
        field: str = "credit amount",
        allow_zero: bool = True,
    ) -> CreditAmount:
        return cls(credit_atoms(value, field=field, allow_zero=allow_zero))

    @classmethod
    def from_atoms(cls, atoms: int) -> CreditAmount:
        return cls(atoms)

    @property
    def decimal(self) -> Decimal:
        return Decimal(credit_decimal(self.atoms))

    def __str__(self) -> str:
        return credit_decimal(self.atoms)

File: typescript/src/credit-amount.ts
import { POSTGRES_BIGINT_MAX } from "./bounds.js";

export const CREDIT_SCALE = 1_000_000n;
export const CREDIT_DECIMAL_PLACES = 6;
export const MAX_CREDIT_ATOMS = POSTGRES_BIGINT_MAX;
export const MAX_WHOLE_CREDITS = MAX_CREDIT_ATOMS / CREDIT_SCALE;

const PLAIN_DECIMAL = /^(0|[1-9][0-9]*)(?:\.([0-9]{1,6}))?$/u;
const CREDIT_AMOUNT_INSTANCES = new WeakSet<object>();

export type CreditInput = string | number | bigint;

function plainDecimalText(value: unknown, field: string): string {
  if (typeof value === "bigint") {
    if (value < 0n) {
      throw new RangeError(`${field} must be non-negative`);
    }
    return value.toString();
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value) || value < 0) {
      throw new TypeError(
        `${field} must be a non-negative safe integer or decimal string`,
      );
    }
    return value.toString();
  }
  if (typeof value === "string") {
    return value;
  }
  throw new TypeError(`${field} must be an integer or plain decimal string`);
}

export function creditAtoms(
  value: unknown,
  options: { readonly field?: string; readonly allowZero?: boolean } = {},
): bigint {
  const field = options.field ?? "credit amount";
  const allowZero = options.allowZero ?? true;
  const text = plainDecimalText(value, field);
  const match = PLAIN_DECIMAL.exec(text);
  if (match === null) {
    throw new RangeError(
      `${field} must be a non-negative plain decimal with at most ${CREDIT_DECIMAL_PLACES} fractional digits`,
    );
  }
  const wholeText = match[1];
  if (wholeText === undefined) {
    throw new RangeError(`${field} is invalid`);
  }
  const whole = BigInt(wholeText);
  if (whole > MAX_WHOLE_CREDITS) {
    throw new RangeError(`${field} exceeds the PostgreSQL bigint atom range`);
  }
  const fractionalText = (match[2] ?? "").padEnd(CREDIT_DECIMAL_PLACES, "0");
  const atoms = whole * CREDIT_SCALE + BigInt(fractionalText || "0");
  if (atoms > MAX_CREDIT_ATOMS) {
    throw new RangeError(`${field} exceeds the PostgreSQL bigint atom range`);
  }
  if (!allowZero && atoms === 0n) {
    throw new RangeError(`${field} must be greater than zero`);
  }
  return atoms;
}

export function creditDecimal(atoms: bigint, field = "credit atoms"): string {
  if (typeof atoms !== "bigint" || atoms < 0n) {
    throw new TypeError(`${field} must be a non-negative bigint`);
  }
  const whole = atoms / CREDIT_SCALE;
  const fractional = atoms % CREDIT_SCALE;
  if (fractional === 0n) {
    return whole.toString();
  }
  const padded = fractional.toString().padStart(CREDIT_DECIMAL_PLACES, "0");
  return `${whole.toString()}.${padded.replace(/0+$/u, "")}`;
}

export function checkedAddAtoms(
  left: bigint,
  right: bigint,
  field = "credit balance",
): bigint {
  if (
    typeof left !== "bigint" ||
    typeof right !== "bigint" ||
    left < 0n ||
    right < 0n ||
    left > MAX_CREDIT_ATOMS ||
    right > MAX_CREDIT_ATOMS
  ) {
    throw new RangeError(
      `${field} operands must be non-negative PostgreSQL bigint values`,
    );
  }
  if (left > MAX_CREDIT_ATOMS - right) {
    throw new RangeError(`${field} exceeds the PostgreSQL bigint atom range`);
  }
  return left + right;
}

export class CreditAmount {
  public readonly atoms: bigint;

  public constructor(atoms: bigint) {
    if (typeof atoms !== "bigint" || atoms < 0n) {
      throw new RangeError("credit atoms must be a non-negative bigint");
    }
    this.atoms = atoms;
    CREDIT_AMOUNT_INSTANCES.add(this);
  }

  public static isCreditAmount(value: unknown): value is CreditAmount {
    return (
      typeof value === "object" &&
      value !== null &&
      CREDIT_AMOUNT_INSTANCES.has(value)
    );
  }

  public static parse(
    value: unknown,
    options: { readonly field?: string; readonly allowZero?: boolean } = {},
  ): CreditAmount {
    return new CreditAmount(creditAtoms(value, options));
  }

  public static fromAtoms(atoms: bigint): CreditAmount {
    return new CreditAmount(atoms);
  }

  public toString(): string {
    return creditDecimal(this.atoms);
  }
}
